### PR TITLE
Misc fixes/improvements

### PR DESCRIPTION
**Pull request by skrawcz**
Fixes/improvees a few things:

1. [Fixes bug where passing in graph_attrs would override all values](https://github.com/DAGWorks-Inc/hamilton/commit/e7021181d94609ad5de50f6dc1140ca74ea31055)
2. Captures more metadata from loaded dataframes.
3. updates package.json for hub.

## Changes
1. graph.py graphviz viz fix. Updates tests
2. updates all the dataframe savers & loaders to pull extra metadata. Updates tests.
3. updates package.json for contrib site.

## How I tested this
 - locally

## Checklist

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/io/utils.py [hunk 1]
```diff
@@ -21,6 +25,39 @@ def get_file_metadata(path: str) -> Dict[str, Any]:
...
+    """
+    return {
+        "rows": len(df),
+        "columns": len(df.columns),
+        "column_names": list(df.columns),
+        "datatypes": [str(t) for t in list(df.dtypes)],  # for serialization purposes
+    }
+
+
+def get_file_and_dataframe_metadata(path: str, df: pd.DataFrame) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/io/utils.py:44]: yep! I'm assuming no one will complain loudly  😬 .

File: hamilton/graph.py [hunk 2]
```diff
@@ -336,7 +336,11 @@ def _get_legend(node_types: Set[str]):
             concentrate="true",
         ),
     )
-    digraph_attr.update(**graphviz_kwargs)
+    for g_key, g_value in graphviz_kwargs.items():
+        if isinstance(g_value, dict):
+            digraph_attr[g_key].update(**g_value)
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/graph.py:366]: it's due to nested dicts.

File: setup.cfg [hunk 3]
```diff
@@ -12,3 +12,6 @@ ignore =
     W503,
     # invalid escape sequence
     W605
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply setup.cfg:14]: yeah maybe that's it. or it's something with tests being a package itself...

File: hamilton/graph.py [hunk 4]
```diff
@@ -51,18 +51,45 @@ def add_dependency(
     """
     if param_name in nodes:
         # validate types match
-        required_node = nodes[param_name]
-        if not types_match(adapter, param_type, required_node.type):
+        required_node: Node = nodes[param_name]
```
Comment (elijahbenizzy) [inline hamilton/graph.py:54]: Type hint necessary? Should be a `dict[str, Node]`

File: hamilton/graph.py [hunk 5]
```diff
@@ -51,18 +51,45 @@ def add_dependency(
...
+            types_are_compatible = types_match(adapter, required_node.type, param_type)
+            if not types_are_compatible:
+                raise ValueError(
+                    f"Error: Two or more functions are requesting {param_name}, but have incompatible types. "
+                    f"{func_name} requires {param_name} to be {param_type}, but found another function(s) "
+                    f"{[f.__name__ for f in required_node.originating_functions]} that minimally require {param_name} "
+                    f"as {required_node.type}. Please fix this by ensuring that all functions requesting {param_name} "
+                    f"have compatible types. If you believe they are equivalent, please reach out to the developers. "
+                    f"Note that, if you have types that are equivalent for your purposes, you can create a "
+                    f"graph adapter that checks the types against each other in a more lenient manner."
```
Comment (elijahbenizzy) [inline hamilton/graph.py:69]: Link to docs here maybe?

File: tests/test_graph.py [hunk 6]
```diff
@@ -108,6 +110,185 @@ def test_add_dependency_strict_node_dependencies():
...
+    b_sig = inspect.signature(tests.resources.compatible_input_types.b)
+    c_sig = inspect.signature(tests.resources.compatible_input_types.c)
+    d_sig = inspect.signature(tests.resources.compatible_input_types.d)
+
+    nodes = {
+        "b": node.Node.from_fn(tests.resources.compatible_input_types.b),
+        "c": node.Node.from_fn(tests.resources.compatible_input_types.c),
+        "d": node.Node.from_fn(tests.resources.compatible_input_types.d),
+    }
+    nodes["b"]._originating_functions = (tests.resources.compatible_input_types.b,)
```
Comment (elijahbenizzy) [inline tests/test_graph.py:250]: `.copy_with` should be able to do this more ergonomically

File: tests/test_graph.py [hunk 7]
```diff
@@ -108,6 +110,185 @@ def test_add_dependency_strict_node_dependencies():
...
+        nodes["d"],
+        "d",
+        nodes,
+        param_name,
+        d_sig.parameters[param_name].annotation,
+        base.SimplePythonDataFrameGraphAdapter(),
+    )
+
+
+def test_add_dependency_input_nodes_compatible_types_order_check():
```
Comment (elijahbenizzy) [inline tests/test_graph.py:239]: Not convinced this test is necessary -- a lot of specific stuff for a strange edge case? Basically copied from the previous? Is there something more specific to test? Otherwise, yeah, comment more -- why would order matter/break it. Could also parameterize (E.G. `reversed`) to share code with the previous.

File: tests/test_hamilton_driver.py [hunk 8]
```diff
@@ -136,7 +136,7 @@ def test_driver_variables_exposes_original_function():
         var.name: var.originating_functions for var in dr.list_available_variables()
     }
     assert originating_functions["b"] == (tests.resources.very_simple_dag.b,)
-    assert originating_functions["a"] is None
```
Comment (elijahbenizzy) [inline tests/test_hamilton_driver.py:139]: Heh this was supposed to be the contract/why I was originally hesitant to make inputs in originating functions

**Review by elijahbenizzy — COMMENTED** [review]
A few nits

File: hamilton/graph.py (hunk 4 above)
Comment (skrawcz) [reply hamilton/graph.py:54]: No, it was useful for pycharm so I could debug it. Otherwise it is a node object, not a dict.

File: hamilton/graph.py (hunk 5 above)
Comment (skrawcz) [reply hamilton/graph.py:69]: No because links might change.

File: tests/test_hamilton_driver.py (hunk 8 above)
Comment (skrawcz) [reply tests/test_hamilton_driver.py:139]: yep that's a good test -- and now we're changing it :)

File: tests/test_graph.py (hunk 7 above)
Comment (skrawcz) [reply tests/test_graph.py:239]: added comment -- I don't see value in refactoring the code because it's pretty straightforward and only has two cases.

File: tests/test_graph.py (hunk 6 above)
Comment (skrawcz) [reply tests/test_graph.py:254]: it doesn't remove lines of code. So I don't feel like it adds much.

**Review by elijahbenizzy — APPROVED** [review]
Looks good to me
